docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,232 @@
+
+2) The runtime attempts to find a binary with the name supplied in `DllImportAttribute`.
+
+    * Discovery of the target binary is complicated and can be influenced by the [`AssemblyLoadContext`](https://docs.microsoft.com/dotnet/api/system.runtime.loader.assemblyloadcontext) and [`NativeLibrary`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary) classes.
+
+3) Once the binary is found and loaded into the runtime, it is queried for the expected export name. The name of the attributed function is used by default but this is configurable by the [`DllImportAttribute.EntryPoint`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.entrypoint) property.
+
+    * This process is also influenced by additional `DllImportAttribute` properties as well as by the underlying platform. For example, the [Win32 API ANSI/UNICODE convention](https://docs.microsoft.com/windows/win32/intl/conventions-for-function-prototypes) on Windows is respected and a(n) `A`/`W` suffix may be appended to the function name if it is not immediately found.
+
+4) The IL Stub is called like any other .NET method.
+
+5) The IL Stub marshals arguments as appropriate and invokes the export via the `calli` instruction.
+
+6) Once the export returns control to the IL Stub the marshaling logic cleans up and ensures any returned data is marshaled back out to the calling function.
+
+### Source Generator Integration
+
+An example of how the previous P/Invoke snippet could be transformed is below. This example is using the proposed API in this document. The Source Generator currently has a limitation that we will ignore at present - the inability to modify user written code. This limitation is ignored to present an optimal user experience, but will be discussed later in case the current limitation becomes a hard requirement.
+
+`Program.cs` (Pre-generated code)
+
+``` CSharp
+/* A */ [DllImportAttribute("Kernel32.dll")]
+/* D */ [GeneratedDllImportAttribute]
+/* B */ extern static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C*/ QueryPerformanceCounter(out count);
+```
+
+Observe point (D), the new attribute. This attribute provides an indication to a Source Generator that the following declaration should be updated.
+
+`Program.cs` (In-memory updated code)
+
+``` CSharp
+/* D */ [GeneratedDllImportAttribute]
+/* B */ partial static bool QueryPerformanceCounter(out long lpPerformanceCount);
+...
+long count;
+/* C */ QueryPerformanceCounter(out count);
+```
+
+During the source generation process the `DllImportAttribute` (A) would be removed. The `GeneratedDllImportAttribute` (D) would remain to provide a metadata indication that the function was auto-generated. Also note that the method declaration has been updated and marked `partial`. The Source Generator would then generate the source for this partial method. The invocation (C) remains unchanged.
+
```

Comment:
There are several problematic points here.

1. I believe that the current plan for source generators will not allow modification of existing source.
2. partial methods must have the return type of void

We should bring this question to the C# language developers and see what they come up for how this should be notated in source.

Reply:
CC. @jaredpar who should be able to ping the right people

Reply:
(1) I discuss further in the proposal.

(2) if unfortunate. I didn't realize that limitation with methods marked `partial`. I just read the doc. Hopefully the C# team can help here.

Reply:
> I believe that the current plan for source generators will not allow modification of existing source.

+100

> partial methods must have the return type of void

That would require a change for sure but I don't think it's a deal breaker.

Reply:
> > I believe that the current plan for source generators will not allow modification of existing source.
> 
> +100

Yep. I hear that is and continues to be a goal. Please see the section on Source Generator constraint mitigation.

> > partial methods must have the return type of void
> 
> That would require a change for sure but I don't think it's a deal breaker.

This would be very helpful to understand what we could get away with here.